### PR TITLE
misc: lint parser.py

### PR DESCRIPTION
**Pull request by superlopuh**
The procrastilinting continues

### CODE REVIEW

File: xdsl/parser.py [hunk 1]
```diff
@@ -85,81 +86,6 @@ def __hash__(self):
         return id(self)
 
 
-@dataclass(frozen=True)
```
Comment (superlopuh) [inline xdsl/parser.py:88]: moved to separate file

File: xdsl/utils/span.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from io import StringIO
+
+
+@dataclass(frozen=True)
```
Comment (superlopuh) [inline xdsl/utils/span.py:7]: unchanged from parser.py

File: xdsl/parser.py [hunk 3]
```diff
@@ -184,52 +110,6 @@ def string_contents(self):
         return ast.literal_eval(self.text)
 
 
-@dataclass(frozen=True)
```
Comment (superlopuh) [inline xdsl/parser.py:187]: also cut/pasted into span.py The two classes reference each other, and I didn't want to change the definitions, so I put them in the same file. Maybe the file should be called input.py. Maybe span_input.py. Happy to discuss :)

File: xdsl/parser.py [hunk 4]
```diff
@@ -1094,6 +977,10 @@ def parse_operation(self) -> Operation:
         # Check for custom op format
         op_name = self.try_parse_bare_id()
         if op_name is not None:
+            assert isinstance(
+                self, XDSLParser
+            ), "Only xDSL format currently supports custom op parsing"
+            assert ret_types is not None, "Return types must be in xDSL format"
```
Comment (webmiche) [inline xdsl/parser.py:983]: Maybe this is github reviews being too local, but I don't quite get the error message here. This sounds like it always needs to be the case, which is not true, right? Is this only on an execution path that can be reached when parsing xDSL syntax?

File: xdsl/parser.py [hunk 5]
```diff
@@ -1293,6 +1180,11 @@ def _parse_builtin_dense_attr(self, _name: Span) -> Attribute | None:
         self.parse_characters(":", err_msg)
         type = self.expect(self.try_parse_type,
                            "Dense attribute must be typed!")
+
+        # Make sure type is AnyTensorType
+        assert isinstance(type, VectorType | TensorType | UnrankedTensorType)
```
Comment (webmiche) [inline xdsl/parser.py:1185]: Why not AnyTensorType? Can you not do isinstance on TypeVars?

I guess this is just one of the downsides when you throw a type system at a language like python.

File: xdsl/parser.py [hunk 6]
```diff
@@ -1399,7 +1291,8 @@ def try_parse_ref_attr(self) -> SymbolRefAttr | None:
         else:
             return None
 
-    def try_parse_builtin_int_attr(self) -> IntegerAttr | None:
+    def try_parse_builtin_int_attr(
+            self) -> IntegerAttr[IntegerType | IndexType] | None:
```
Comment (webmiche) [inline xdsl/parser.py:1295]: Is there a way to differentiate among the two to get stronger type guarantees? Maybe through two more functions, one that returns an IntegerAttr of IntegerType, the other of IndexType. And then, when using those in custom parsing, you should get better type guarantees...

Maybe this is out of scope for this PR though: Curious to hear your thoughts :)

File: xdsl/parser.py [hunk 7]
```diff
@@ -1429,7 +1327,8 @@ def try_parse_builtin_float_attr(self) -> FloatAttr | None:
                     "Float attribute must be typed with a float type!")
             return FloatAttr(float(value.text), type)
 
-    def try_parse_builtin_boolean_attr(self) -> IntegerAttr | None:
+    def try_parse_builtin_boolean_attr(
```
Comment (webmiche) [inline xdsl/parser.py:1330]: Can builtin bool attr actually be of IndexType? IMO, we should only allow i1, which is an IntegerType, right?

**Review by webmiche — COMMENTED** [review]
I like the split into two files. I am now wondering if we should also mention in the filename or in some comment at the start of the file that this is something mainly used in the parser.

Maybe this is for a broader discussion on include dependencies and repo structure though :)

File: xdsl/parser.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/parser.py:983]: Yeah that's the case. both of the asserts are just putting in code the implicit assumptions. return types are printed before the op only in xDSL syntax, in MLIR they're at the end of the line.

Comment (webmiche) [reply xdsl/parser.py:983]: Okay, works for me :+1:

File: xdsl/parser.py (hunk 5 above)
Comment (webmiche) [reply xdsl/parser.py:1185]: I would prefer isa, feels like more reuse to me. Not a strong opinion though...

File: xdsl/parser.py (hunk 7 above)
Comment (webmiche) [reply xdsl/parser.py:1330]: I don't mind. This one really isn't that big, but I see that this is somewhat out of scope. Your call...

**Review by webmiche — APPROVED** [review]
LGTM

File: xdsl/parser.py (hunk 5 above)
Comment (superlopuh) [reply xdsl/parser.py:1185]: - [ ] use isa

File: xdsl/parser.py [hunk 8]
```diff
@@ -932,9 +813,11 @@ def unimplemented() -> ParametrizedAttribute:
     def parse_complex_attrs(self):
         self.raise_error("ComplexType is unimplemented!")
 
-    def parse_memref_attrs(self) -> MemRefType | UnrankedMemrefType:
+    def parse_memref_attrs(
+            self) -> MemRefType[Attribute] | UnrankedMemrefType[Attribute]:
         dims = self._parse_tensor_or_memref_dims()
         type = self.try_parse_type()
+        assert type is not None
```
Comment (AntonLydike) [inline xdsl/parser.py:820]: Maybe switch to `parser.expect` here?

**Review by AntonLydike — APPROVED** [review]
Looks good to me, other than a small nitpick. Thanks for doing this work!

File: xdsl/parser.py [hunk 9]
```diff
@@ -1411,9 +1304,14 @@ def try_parse_builtin_int_attr(self) -> IntegerAttr | None:
...
             type = self._parse_attribute_type()
+
+            if not isinstance(type, IntegerType | IndexType):
+                self.raise_error(
+                    f"Expected IntegerType | IndexType, got {type}")
+
             return IntegerAttr.from_params(int(value.text), type)
 
-    def try_parse_builtin_float_attr(self) -> FloatAttr | None:
+    def try_parse_builtin_float_attr(self) -> FloatAttr[AnyFloat] | None:
```
Comment (math-fehr) [inline xdsl/parser.py:1314]:
```suggestion
    def try_parse_builtin_float_attr(self) -> AnyFloatAttr | None:
```
?

File: xdsl/parser.py [hunk 10]
```diff
@@ -1448,7 +1347,7 @@ def try_parse_builtin_str_attr(self):
                 self.raise_error("Invalid string literal")
             return StringAttr(literal.string_contents)
 
-    def try_parse_builtin_arr_attr(self) -> ArrayAttr | None:
+    def try_parse_builtin_arr_attr(self) -> ArrayAttr[Attribute] | None:
```
Comment (math-fehr) [inline xdsl/parser.py:1350]:
```suggestion
    def try_parse_builtin_arr_attr(self) -> AnyArrayAttr | None:
```

**Review by math-fehr — APPROVED** [review]
Nice!
I added some nit comments

File: xdsl/parser.py (hunk 8 above)
Comment (superlopuh) [reply xdsl/parser.py:820]: - [ ] switch to parser.expect

File: xdsl/parser.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/parser.py:88]: - [ ] rename to lexer.py

File: xdsl/parser.py [hunk 11]
```diff
@@ -746,14 +628,13 @@ def parse_list_of(self,
...
-        items = list()
         first_item = try_parse()
         if first_item is None:
             if allow_empty:
-                return items
+                return []
             self.raise_error(error_msg)
-
-        items.append(first_item)
+        else:
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/parser.py:636]: felt appropriate, but thinking more I guess it's not necessary, I'll move it out

- [ ] move it out

File: xdsl/parser.py [hunk 12]
```diff
@@ -1094,6 +977,10 @@ def parse_operation(self) -> Operation:
         # Check for custom op format
         op_name = self.try_parse_bare_id()
         if op_name is not None:
+            assert isinstance(
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/parser.py:980]: I don't think so, this bit of code should not be reachable. Custom parsing can only happen with the xDSL parser, this is a reification of an API contract, so assert feels more appropriate. How would you recover from an exception raised here?

File: xdsl/parser.py (hunk 9 above)
Comment (superlopuh) [reply xdsl/parser.py:1314]: - [ ] AnyFloatAttr

File: xdsl/parser.py (hunk 10 above)
Comment (superlopuh) [reply xdsl/parser.py:1350]: - [ ] AnyArrayAttr

**Review by georgebisbas — APPROVED** [review]
